pymongo/uri_parser.py

```diff
@@ -34,26 +40,49 @@
 SRV_SCHEME_LEN = len(SRV_SCHEME)
 DEFAULT_PORT = 27017
 
+# List of sub-delimiters as defined in RFC 3986.
+SUBDELIMS = ["!", "$", "&", "'", "(", ")", "*", "+", ",", ";", "="]
+
 
 def parse_userinfo(userinfo):
     """Validates the format of user information in a MongoDB URI.
-    Reserved characters like ':', '/', '+' and '@' must be escaped
-    following RFC 3986.
+    Reserved characters that are gen-delimiters (":", "/", "?", "#", "[",
+    "]", "@") as per RFC 3986 must be escaped.
 
     Returns a 2-tuple containing the unescaped username followed
     by the unescaped password.
 
     :Paramaters:
         - `userinfo`: A string of the form <username>:<password>
+
+    .. versionchanged:: 4.0
+       Reserved characters that are sub-delimiters ("!", "$", "&", "'",
+       "(", ")", "*", "+", ",", ";", "=") as per RFC 3986 need not be
+       escaped.
     """
     if '@' in userinfo or userinfo.count(':') > 1:
         raise InvalidURI("Username and password must be escaped according to "
                          "RFC 3986, use urllib.parse.quote_plus")
+        # If we cannot round-trip a value, it is not properly URI encoded.
+        # This checks that we can take a value, quote it, and then unquote
+        # that quoted string and arrive at the same value.
+        # We take in the unquoted user and passwd strings and
+        # strip subdelimeters and then compare the original unquoted value
+        # without subdelimeters to the resulting string after round-tripping.
+        # One round trip means to quote once and then unquote.
+
+    unquoted_value_no_sdelims = "".join(
+        [ch for ch in userinfo if ch not in SUBDELIMS])
+    quoted_value = quote_plus(unquoted_value_no_sdelims)
+    if unquoted_value_no_sdelims != unquote_plus(quoted_value):
```

Comment:
We should change this to use quote instead of quote_plus, then we can also fit it on one line:
```
if no_subdelims != unquote(quote(no_subdelims)):
```

Reply:
Done.